pkg/clusterstatus/etcd/etcd.go

```diff
@@ -0,0 +1,212 @@
+/*
+Copyright 2019 The KubeOne Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+
+	kubeoneapi "github.com/kubermatic/kubeone/pkg/apis/kubeone"
+	"github.com/kubermatic/kubeone/pkg/httptunnel"
+	"github.com/kubermatic/kubeone/pkg/state"
+)
+
+const (
+	healthEndpoint  = "https://%s:2379/health"
+	membersEndpoint = "https://127.0.0.1:2379/v2/members"
+)
+
+// Status describes status of the etcd cluster
+type Status struct {
+	Health bool `json:"health,omitempty"`
+	Member bool `json:"member,omitempty"`
+}
+
+type healthRaw struct {
+	Health string `json:"health"`
+}
+
+type membersListRaw struct {
+	Members []struct {
+		ID         string   `json:"id,omitempty"`
+		Name       string   `json:"name,omitempty"`
+		PeerURLs   []string `json:"peerURLs,omitempty"`
+		ClientURLs []string `json:"clientURLs,omitempty"`
+	} `json:"members"`
+}
+
+type HTTPDoer interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+// EtcdStatus analyzes health of an etcd cluster
+func GetStatus(s *state.State, node kubeoneapi.HostConfig) (*Status, error) {
+	tlsCfg, err := loadTLSConfig(s)
+	if err != nil {
+		return nil, err
+	}
+	client, err := httptunnel.NewHTTPTunnel(s, tlsCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	m, err := membersList(client)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check etcd member health
+	healthStr, err := memberHealth(client, node.PrivateAddress)
+	if err != nil {
+		return nil, err
+	}
+	health, err := strconv.ParseBool(healthStr.Health)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check etcd membership
+	var member bool
+	for _, mem := range m.Members {
+		if mem.Name == node.Hostname {
+			member = true
+			break
+		}
+	}
+
+	return &Status{
+		Health: health,
+		Member: member,
+	}, nil
+}
+
+// memberHealth returns health for a requested etcd member
+func memberHealth(t *httptunnel.HTTPTunnel, nodeAddress string) (*healthRaw, error) {
```

Comment:
same as in pkg/clusterstatus/apiserver/apiserver.go, tunneler -> http.Client or even better — an interface.